### PR TITLE
Handle CGINCLUDE blocks anywhere in file

### LINKED ISSUE #188: CGINCLUDE blocks can be anywhere
The ShaderLab parser makes an incorrect assumption about the location of `CGINCLUDE` (and `HLSLINCLUDE` and `GLSLINCLUDE`) blocks, which can cause syntax errors. The `CGINCLUDE` block can be placed anywhere (`CGPROGRAM` et al must still be placed inside a `Pass` block).

Compiling a shader can be thought of as a multi-pass operation. The first pass removes all `CGINCLUDE` and `CGPROGRAM` blocks, compiles the shader and replaces the `CGPROGRAM` block with a `GpuProgramID <num>` command, which is a lookup to identify the compiled code. The compiled code is generated from prepending any `CGINCLUDE` blocks to the start of each `CGPROGRAM` block.

### PR DESCRIPTION
**Pull request by citizenmatt**
Fixes #188. Builds upon PR #189

### CODE REVIEW

**Comment by citizenmatt** [comment]
Possibly should have pushed this onto #189, but perhaps this makes it easier to review?

File: resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi [hunk 1]
```diff
@@ -1812,26 +1813,61 @@ bindValue
...
+  includeBlock
+  (errorElement | WHITESPACE | NEW_LINE)
+;
+
+// This rule isn't called by anything in the generated parser, but is used
+// by the preprocessor. When a shader is compiled, the CG_INCLUDE blocks et al
+// are pre-processed and removed. This means they can appear anywhere in the
+// file. We treat them like pre-processor directive tokens, parse them separately
+// filter them, and then reinsert them after normal parsing. Also note that we
+// have to handl whitespace here, as the pre-processor doesn't use the filtering
```
Comment (IvanPashchenko) [inline resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi]: > handl

a typo here :)

File: resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi [hunk 2]
```diff
@@ -1812,26 +1813,61 @@ bindValue
...
+// lexer. As it happens, this construct doesn't care about whitespace
+// (Technically, this is also true of CG_PROGRAM et al; they are pre-processed
+// and replaced with a GpuProgramId command that takes the index of the
+// compiled shader code. But they can only appear inside a Pass, so we can
```
Comment (IvanPashchenko) [inline resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi]: > inside a Pass

surface shader are written inside a `SubShader` section
https://docs.unity3d.com/Manual/SL-SurfaceShaderExamples.html

Comment (citizenmatt) [inline resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi]: Oh, er, yeah, I knew that :)

**Pull request merged**